### PR TITLE
Rework logger class and add unit testing for it

### LINKED ISSUE #70: Unit Test Logger

### PR DESCRIPTION
**Pull request by DrewBrandt**
This makes some pretty major reworks to the way the PSRAM works. First, it now has a BASIC file system that is based on FAT. Second, it supports compressing data so that high-refresh-rate flights don't run into memory issues on the chip.

See [The README](https://github.com/Terrapin-Rocket-Team/Multi-Mission-Flight-Software/blob/unit-test-logger/src/RecordData/README.md) for a hopefully full explanation of how it all works.

I've mentioned that I'm not confident that these systems work, but I should clarify that. I don't think they don't work. I simply think that more edge cases and use cases need to be tested, and it needs hardware testing. I am confident that the default use-case works based on the flight and on the unit tests that I wrote. I simply think I need to write more of them at some point.

Closes #70 and #81

### CODE REVIEW

**Comment by varun-un** [comment]
how will this scale if we want to have 2 PSRAM chips, to help deal with the PSRAM memory limits that Payload may hit? Could change the `Logger` class to use an array of PSRAMs but that would need some changing. In addition, I don't know if the line:
```cpp
extern mmfs::PSRAM *psram;
```
may cause issues with multiple psrams.

This may also be jumping the gun so @mallamacimj @DrewBrandt, is this a realistic option for now, or are we not gonna have to worry about using multiple PSRAMs?

**Comment by mallamacimj** [comment]
We are prob fine with 1 psram for now. We can change it later if it becomes an issue. But for the foreseeable future we are just going to use 1.

**Comment by DrewBrandt** [comment]
Psram is treated as raw memory and I don't know but assume they would be contiguous. So the psram size would just be bigger but we would still only be using "one" psram

File: src/RecordData/DataFormatter.cpp [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    int strSize = 0;
+    for (int i = 0; i < obj->getNumPackedDataPoints() && destLen > 0; i++)
+    {
+        const void *dataPtr = (uint8_t *)data + dataOffset;
+        dataOffset += obj->PackedTypeToSize(obj->getPackedOrder()[i]);
+        int printedSize = 0;
+        switch (obj->getPackedOrder()[i])
+        {
+        case FLOAT:
+            printedSize = snprintf(dest + strSize, destLen, "%.3f,", *(float *)dataPtr);
```
Comment (varun-un) [inline src/RecordData/DataFormatter.cpp:98]: Can you use the return value like this?

"*The snprintf() function returns the number of characters that would have been written had n been sufficiently large, not counting the terminating null character, or a negative value if an encoding error occurred*"

File: src/RecordData/psram.h [hunk 2]
```diff
@@ -1,156 +1,50 @@
...
+        friend class PSRAMFile;
 
-    /**
-     * Print a string to the PSRAM followed by a newline with an option to write to the top.
-     * @param data The string to be printed.
-     * @param writeToTop If true, write to the top of the memory. If false, write to the bottom in reverse.
-     */
-    void println(const char *data, bool writeToTop);
+        inline int min(int a, int b) { return a < b ? a : b; }
+        inline int max(int a, int b) { return a > b ? a : b; }
```
Comment (varun-un) [inline src/RecordData/psram.h:17]: doesn't Arduino and/or C++11 have this built-in? Is this just to be platform-agnostic

**Review by varun-un — COMMENTED** [review]
looks good for the most part, though I will say that the system is a bit more complex now (whether that's a good thing or bad thing I'd say we don't know yet).

This code does also use `new` a few times, so we should make sure all those operations are safe for failure modes

File: src/RecordData/psram.cpp [hunk 3]
```diff
@@ -1,218 +1,215 @@
...
-            if (writeToTop)
-            {
-                *cursorStart = data[i];
-                cursorStart++;
-            }
-            else
+            // Change this cluster's next cluster to the new one, preserving the previous cluster
+            clusterMap[file.eofCluster] = (clusterMap[file.eofCluster] & 0xFF000000) | (nextCluster << 16) | file.id;
+            // set the next cluster's previous cluster to the current cluster
+            clusterMap[nextCluster] = (file.eofCluster << 24) | (0xFF << 16) | file.id; // there is no next cluster
```
[the remark this answers is not in the document]
Comment (DrewBrandt) [reply src/RecordData/psram.cpp:70]: yes

File: src/RecordData/PSRAMFile.h [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+#include "../Constants.h"
+
+namespace mmfs
+{
+    inline const uint8_t F_READ = 0b001;
+    inline const uint8_t F_WRITE = 0b010;
+    inline const uint8_t F_APPEND = 0b100;
+
+    inline const uint8_t F_SEEK_SET = 1;
+    // inline const uint8_t F_SEEK_CUR = 2; // not implemented
```
[the remark this answers is not in the document]
Comment (DrewBrandt) [reply src/RecordData/PSRAMFile.h:14]: mostly just imitating a real file system lol. Basically if you don't set append, write will open at the start of the file otherwise the end

File: src/RecordData/psram.cpp [hunk 5]
```diff
@@ -1,218 +1,215 @@
...
 
-void PSRAM::println(const char *data, bool writeToTop)
-{
-    if (ready)
-    {
-        print(data, writeToTop);
-        print("\n", writeToTop);
-    }
-}
+            if (nextCluster == (uint8_t)-1) // No more free clusters (out of space)
```
[the remark this answers is not in the document]
Comment (DrewBrandt) [reply src/RecordData/psram.cpp:64]: uhhh yeah. I need to figure out a way to allow more clusters so this isn't an issue, I just haven't yet.

File: src/RecordData/PSRAMFile.h [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    class PSRAMFile
+    {
+
+        friend class PSRAM;
+
+    private:
+        uintptr_t endOfFile = 0;                   // address of the last byte written
+        uintptr_t cursor = 0;                      // address of current location, offset by PSRAM_STARTING_ADDRESS
+        uint8_t clusterCursor = 0;                 // similar to cursor, but used for reading one cluster at a time, index of the current cluster
+        char *name = new char[MAX_FILE_NAME_SIZE]; // file name
```
[the remark this answers is not in the document]
Comment (DrewBrandt) [reply src/RecordData/PSRAMFile.h:25]: fixed

File: src/RecordData/PSRAMFile.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+}
+
+void PSRAMFile::readTo(char *data, char endChar)
+{
+    // unimplemented
+}
+
+void PSRAMFile::setName(const char *name)
+{
+    strncpy(this->name, name, MAX_FILE_NAME_SIZE);
```
[the remark this answers is not in the document]
Comment (DrewBrandt) [reply src/RecordData/PSRAMFile.cpp:50]: ok fixed

File: src/RecordData/DataFormatter.cpp (hunk 1 above)
Comment (DrewBrandt) [reply src/RecordData/DataFormatter.cpp:98]: yes it's the number readable of characters printed. If destLen runs out of space, the next iteration of the for loop will catch that and stop any further print attempts

File: src/RecordData/psram.h (hunk 2 above)
Comment (DrewBrandt) [reply src/RecordData/psram.h:17]: yeah exactly

Comment (DrewBrandt) [reply src/RecordData/psram.h:17]: It was needed for testing on native but instead of making this file reliant on arduino.h I just put these two small functions here

**Comment by DrewBrandt** [comment]
I agree the system is much more complicated now. I also don't know if that's a good or bad thing. The biggest limitation is that we can only have <= 255 clusters which means each file takes at least 32KB of space. I could pretty easily expand the cluster map to allow 16 bit addresses, but that doubles the size of that map to about 2KB in ram. I could also just use the first cluster as PSRAM overhead space and make that change, but idk.
